core/include/scipp/core/element/bin.h

```diff
@@ -0,0 +1,151 @@
+                 index *= scipp::size(edges) - 1;
+                 index += (it == edges.begin() || it == edges.end())
+                              ? -1
+                              : --it - edges.begin();
+               }};
+
+static constexpr auto groups_to_map = overloaded{
+    element::arg_list<span<const int64_t>, span<const int32_t>,
+                      span<const std::string>>,
+    transform_flags::expect_no_variance_arg<0>,
+    [](const units::Unit &u) { return u; },
+    [](const auto &groups) {
+      std::unordered_map<typename std::decay_t<decltype(groups)>::value_type,
+                         scipp::index>
+          index;
+      scipp::index current = 0;
+      for (const auto &item : groups)
+        index[item] = current++;
+      if (scipp::size(groups) != scipp::size(index))
+        throw std::runtime_error("Duplicate group labels.");
+      return index;
+    }};
+
+template <class T>
+using update_indices_by_grouping_arg =
+    std::tuple<scipp::index, T, std::unordered_map<T, scipp::index>>;
+
+static constexpr auto update_indices_by_grouping =
+    overloaded{element::arg_list<update_indices_by_grouping_arg<int64_t>,
+                                 update_indices_by_grouping_arg<int32_t>,
+                                 update_indices_by_grouping_arg<std::string>>,
+               [](units::Unit &indices, const units::Unit &coord,
+                  const units::Unit &groups) {
+                 expect::equals(coord, groups);
+                 expect::equals(indices, units::one);
+               },
+               [](auto &index, const auto &x, const auto &groups) {
+                 if (index == -1)
+                   return;
+                 const auto it = groups.find(x);
+                 index *= scipp::size(groups);
+                 index += it == groups.end() ? -1 : it->second;
+               }};
+
+static constexpr auto update_indices_from_existing =
+    overloaded{element::arg_list<scipp::index>,
+               [](units::Unit &, const units::Unit &, const units::Unit &) {},
+               [](auto &index, const auto bin_index, const auto nbin) {
+                 if (index == -1)
+                   return;
+                 index *= nbin;
+                 index += bin_index;
+               }};
+
+// - Each span is covers an *input* bin.
+// - `offsets` Start indices of the output bins
+// - `bin_indices` Target output bin index (within input bin)
+template <class T>
+using bin_arg = std::tuple<span<T>, span<const scipp::index>, span<const T>,
+                           span<const scipp::index>>;
+static constexpr auto bin = overloaded{
+    element::arg_list<bin_arg<double>, bin_arg<float>, bin_arg<int64_t>,
+                      bin_arg<int32_t>, bin_arg<bool>, bin_arg<Eigen::Vector3d>,
+                      bin_arg<std::string>>,
+    transform_flags::expect_in_variance_if_out_variance,
+    [](units::Unit &binned, const units::Unit &, const units::Unit &data,
```

Comment:
Question for my own understanding.

I've noticed that all of the overloads have a default implementation where every input is a unit. Is this so the same function can be used for unit conversion side by side with the data conversion?

Reply:
Yes, `transform` applies the functor to:
1. The units of all the input variables.
2. To all the elements (potentially transposed/broadcast) of all the input variables.

In cases where `Unit` supports the some operations as the data, e.g., `+` we can use a generic lambda for both. In other cases we provide an overload for `Unit`.